fix: Fix the vertical size calculation & the zoom on layer disable

oscarshaitan opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/hm21/pro_image_editor/blob/stable/CONTRIBUTING.md#style-guides
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #465 #467


## What is the new behavior?
refactor: replace getMinimumSize with getValidSizeOrDefault in various editors
fix: add `forceIgnoreGestures` to allow zoom gestures when a layer is disable

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

oscarshaitan committed:
feat: update version to 9.9.1 and add changelog entry for main editor size fix

refactor: replace getMinimumSize with getValidSizeOrDefault in various editors

oscarshaitan committed:
fix(main-editor): enhance layout with SafeArea for better UI responsiveness

oscarshaitan committed:
fix(main-editor): correct formatting in ProImageEditorState widget

oscarshaitan committed:
feat(layer-interaction): add forceIgnoreGestures parameter to control gesture interactions

oscarshaitan committed:
fix(main-editor): update changelog for main editor size calculation and zoom gesture issue

oscarshaitan committed:
remove extra SafeArea in PaintEditor

oscarshaitan committed:
Merge branch 'stable' of https://github.com/SymTerra/pro_image_editor into fix/editor-size-calculation

# Conflicts:
#	CHANGELOG.md
#	pubspec.yaml

hm21 commented on lib/shared/widgets/layer/interaction_helper/layer_interaction_helper_widget.dart:
```diff
@@ -175,7 +183,8 @@ class _LayerInteractionHelperWidgetState
     return TooltipVisibility(
       visible: layerInteraction.style.showTooltips,
-      child: DeferPointer(
+      child: IgnorePointer(
```
The `DeferPointer` widget is everywhere mandatory, without it, gesture detection won't work correctly anymore. For example, if you draw a long vertical line which you later rotate 45deg and then try to select it at the ends, it won’t work without `DeferPointer`.

<img src="https://github.com/user-attachments/assets/fe8ef340-89a8-41d1-b485-a92ce064ea94" width="250" />

oscarshaitan replied:
ok Im going to roll back that and add simply early return
```dart
 if (widget.forceIgnoreGestures) {
      return IgnorePointer(
          ignoring: widget.forceIgnoreGestures, child: widget.child);
    }
```

oscarshaitan replied:
that will mix both solutions  preserving the behaviour you mention but also allowing to totally disable the layer

hm21 commented on lib/designs/grounded/widgets/bottombar/grounded_filter_bar.dart:
```diff
@@ -1,10 +1,11 @@
 import 'package:flutter/material.dart';
-
+import 'package:pro_image_editor/core/utils/size_utils.dart';
```
Please update everywhere the import to use the relative path style: `import '/core/utils/size_utils.dart';` for consistency.

hm21 commented on lib/features/main_editor/main_editor.dart:
```diff
@@ -1311,23 +1311,22 @@ class ProImageEditorState extends State<ProImageEditor>
-                              .style.subEditorPage.borderRadius,
-                        ),
-                        child: page,
+                        child: Container(
```
This part can be removed, it only eliminates a comma without making any meaningful change.

hm21 reviewed (changes requested):
Thanks for creating the new PR! The main issue is the removal of `DeferPointer`, unfortunately that’s not possible, and I’ve explained why in the review.

As for the other changes, like renaming `getValidSizeOrDefault` and removing `SafeArea`, I can merge them once you've addressed the review comments there. 

Thanks for your collaboration! I appreciate your help on this.

oscarshaitan committed:
Pr review

oscarshaitan committed:
Pr review

hm21 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## 9.9.3
+- **FIX**(Main-Editor & Paint-Editor): Fix the main editor size calculation to reduce layers shiftting and fix the disable layers not allowign zoom gestures on main editor
```
```suggestion
- **FIX**(Layers): Corrected size calculation to prevent layer shifting.
- **FIX**(Main-Editor): Fixed an issue where disabled layers blocked zoom gestures.
```

hm21 commented on lib/features/blur_editor/blur_editor.dart:
```diff
@@ -5,6 +5,7 @@ import 'dart:async';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
+import 'package:pro_image_editor/core/utils/size_utils.dart';
```
```suggestion
import '/core/utils/size_utils.dart';
```

hm21 commented on lib/features/filter_editor/filter_editor.dart:
```diff
@@ -4,6 +4,7 @@ import 'dart:async';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
+import 'package:pro_image_editor/core/utils/size_utils.dart';
```
```suggestion
import '/core/utils/size_utils.dart';
```

hm21 commented on lib/features/main_editor/widgets/main_editor_layers.dart:
```diff
@@ -1,5 +1,6 @@
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
+import 'package:pro_image_editor/core/utils/size_utils.dart';
```
```suggestion
import '/core/utils/size_utils.dart';
```

hm21 commented on lib/features/tune_editor/tune_editor.dart:
```diff
@@ -4,6 +4,7 @@ import 'dart:async';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
+import 'package:pro_image_editor/core/utils/size_utils.dart';
```
```suggestion
import '/core/utils/size_utils.dart';
```

hm21 commented on lib/shared/widgets/layer/layer_widget.dart:
```diff
@@ -298,6 +298,8 @@ class _LayerWidgetState extends State<LayerWidget>
       callbacks: callbacks,
       selected: widget.selected,
       onEditLayer: widget.onEditTap,
+      forceIgnoreGestures:
```
I'm not sure if that's going to be a problem. But if we use `IgnorePointer` every callback below, it might stop working too. I'm about 90% sure there won't be any issues with custom user logic, but i didn't test it. 
You mentioned before the only problem is with the double-tap interaction, right? What do you think about this approach: 
```suggestion
forceIgnoreGestures:
    !(interaction.enableSelection || interaction.enableEdit) &&
        configs.mainEditor.enableZoom &&
        configs.mainEditor.enableDoubleTapZoom,
```

If you're confident that no necessary callbacks will be blocked, you can ignore it, I'll still go ahead and merge it. I'll leave the decision up to you whether we should extend it or not.

oscarshaitan replied:
I didn't include those because the my idea was if layer interactions are disable allow everything else in the main editor like zoom on any gesture.

hm21 reviewed (changes requested):
Sorry, my bad, I didn't explain it clearly enough. I'd prefer if you could rename every import. This time, I’ve added a review comment to each one. Also, I’d appreciate it if we could separate the changelog messages.

I’ve added a small suggestion for `forceIgnoreGestures`, but feel free to ignore it if you think it’s not necessary.

Once that's done, I’ll go ahead and merge it right away. Thanks again for all the work and time you've put into improving the image editor!

oscarshaitan committed:
Update lib/features/blur_editor/blur_editor.dart

Co-authored-by: Alex Frei <40503456+hm21@users.noreply.github.com>

oscarshaitan committed:
Update lib/features/filter_editor/filter_editor.dart

Co-authored-by: Alex Frei <40503456+hm21@users.noreply.github.com>

oscarshaitan committed:
Update lib/features/main_editor/widgets/main_editor_layers.dart

Co-authored-by: Alex Frei <40503456+hm21@users.noreply.github.com>

oscarshaitan committed:
Update lib/features/tune_editor/tune_editor.dart

Co-authored-by: Alex Frei <40503456+hm21@users.noreply.github.com>

oscarshaitan committed:
Update CHANGELOG.md

Co-authored-by: Alex Frei <40503456+hm21@users.noreply.github.com>

oscarshaitan commented:
> Sorry, my bad, I didn't explain it clearly enough. I'd prefer if you could rename every import. This time, I’ve added a review comment to each one. Also, I’d appreciate it if we could separate the changelog messages.
> 
> I’ve added a small suggestion for `forceIgnoreGestures`, but feel free to ignore it if you think it’s not necessary.
> 
> Once that's done, I’ll go ahead and merge it right away. Thanks again for all the work and time you've put into improving the image editor!

for sure this library is awesome and is helping me build one hell of a cluster feature, so ill try all that I can do to make it better

oscarshaitan committed:
fixing imports

oscarshaitan committed:
fixing imports
